Add dev sandbox to the repo

gribnoysup opened the pull request:
A possible solution for https://github.com/styled-components/styled-components/issues/1238

run `npm run dev` and you should see something like this on `localhost:8080`:

![image](https://user-images.githubusercontent.com/5036933/31819165-20801692-b5a4-11e7-97c5-51eacbc51e3c.png)

philpl commented on package.json:
```diff
@@ -28,7 +28,7 @@
     "prepublish": "npm run build",
     "lint-staged": "lint-staged",
-    "dev": "babel-node example/devServer.js"
+    "dev": "webpack-dev-server --config webpack.config.dev.js"
```
I think we should make this completely separate, i.e. put a package.json into the sandbox folder and so on :smile: Should help us to play around with it a little more

cc @mxstbr

philpl replied:
Also, we don't have to do it in this PR, but the sandbox should have some kind of SSR setup

gribnoysup replied:
> should have some kind of SSR setup

A great idea, but yeah, I'm not sure that I'll be able to do this in this PR

mxstbr replied:
Don't worry about the SSR stuff, let's ship this as a first iteration.

philpl commented on .eslintignore:
```diff
@@ -3,5 +3,6 @@ src/vendor/**/*.js
 src/test/utils.js
 *.test.js
 example/
+sandbox/
```
Actually, I think it's worth it to run prettier and eslint on the sandbox JS files

gribnoysup replied:
@philpl do I also need to add flow annotation to the sandbox, or this eslint rule could be ignored?

philpl replied:
@gribnoysup oh good point! no, I don't think we'll need Flow on the sandbox. Maybe it's worth then to separate the prettier/eslint setup a little

gribnoysup commented:
I updated the CONTRIBUTING.md, but English is not my native language, so I'm not sure that it is any good. Could you please check it thoroughly? :)

philpl commented:
@mxstbr @gribnoysup I think this is perfect as a first step! Thanks ❤️ 

We can keep working on it to add more functionality later on

gribnoysup commented:
A little update:

It seems like `webpack-dev-middleware` supports SSR, although documentation states that this feature is experimental. I could try to work on it, I think I have a general idea how to do this.

I'm also almost done with separating sandbox from the main package into its own folder as @philpl suggested, but there are a few things that I wanted to get your advice on.

I'm in a middle of relocation process, so I'll probably won't be able to continue with this for a few days 😅 I see that you approved this PR, so if you want, you can merge it as is, and I'll create a separate one with additional updates

mxstbr commented:
Let's merge this once the sandbox is separated so folks who `npm i styled-components` don't have to download all of our dependencies too.

gribnoysup commented:
Okay, so I had a very productive flight and added an SSR support, need to add a few minor fixes and will push to this branch

gribnoysup commented:
@philpl @mxstbr 

I added SSR support and then I figured out that unlike `webpack-dev-server` `webpack-dev-middleware` doesn't support live reload, so I had to rebuild it specifically for this sandbox 😅 

I didn't separate eslint config for the sandbox because I thought that if you are using it to fiddle with `styled-components`, it is not a bad idea to see eslint errors for the source you are working with right away, but maybe I am wrong on this one, what do you think? I can change this and throw something like `eslint-create-react-app` instead of using global lobrary config

Also I didn't add flow typings, instead adding `eslint-disable` to supress this rule, but maybe it is not a bad idea to cover at least core parts of the sandbox (everything but `App.js` basically) later. Unfortunately I didn't have time for this right now

gribnoysup commented:
If you think that this PR is too big, I can revert some stuff if you want. I just a little carried away with this livereload SSR stuff 😸

mxstbr commented:
No it's not too big at all, just gotta test it locally!

mxstbr commented:
What on earth, I merged `master` into  this branch and now it's showing the static caching fixes in the diff 🤔

mxstbr commented:
Hmm, i'm now getting some errors here:

<img width="475" alt="screen shot 2017-10-24 at 4 00 13 pm" src="https://user-images.githubusercontent.com/7525670/31947374-77dfa3f2-b8d4-11e7-8bb5-9f3098e03a51.png">

Were these there before or did I just mess everything up with the merge?

mxstbr commented:
Ahhhh I fucked everything up! :scream: @gribnoysup can you do a

```
git push origin issues/1238 --force
```

So that we go back to your original branch state? Sorry about the troubles, I totally messed up the git merge here.

gribnoysup commented:
@mxstbr 

Hmmm, interesting! I actually think that everything is okay with the errors in the console 😄 

There seem to be minor errors in the code style, that are not captured with `npm run lint` or fixed on the pre-commit hook for some reason, but are in the code and sandbox sees them:

https://github.com/styled-components/styled-components/blob/master/src/constructors/constructWithOptions.js#L6

https://github.com/styled-components/styled-components/blob/master/src/utils/create-broadcast.js#L22

mxstbr commented:
Yeah but I still messed up the branch, can you do the force push? Thanks!

gribnoysup commented:
Done!

gribnoysup commented:
Rebased the branch on the master, no conflicts now

philpl commented:
@gribnoysup awesome work 😍 I’ll take another look this evening. Let’s get this merged!

philpl commented on CONTRIBUTING.md:
```diff
@@ -30,7 +30,30 @@ Please also give the code of conduct a read.
 
 ## How do I set up the project?
 
-Run `npm install` and edit code in the `src/` folder. It's luckily very simple! :wink:
```
We shouldn't remove this :wink:

Instead can we add another title: "How do I verify and test my changes?".
There we can not only talk about the sandbox, but also about the tests

philpl commented on package.json:
```diff
@@ -94,6 +93,7 @@
     "enzyme": "^2.8.2",
     "eslint": "^3.15.0",
     "eslint-config-airbnb": "^13.0.0",
+    "eslint-import-resolver-webpack": "^0.8.3",
```
Is this supposed to be in the sub-folder?

gribnoysup replied:
As the sandbox just extends main `styled-components` config, I left this in the main `package.json` on purpose. Otherwise when you `npm install` the sandbox there are a lot of warnings about missing `peerDeps`. Should I move this to the sandbox anyway?

philpl replied:
ah, yea, that makes sense :+1: keep it then :smile:

philpl commented on sandbox/index.js:
```diff
@@ -0,0 +1,63 @@
+
+const webpackDevMiddleware = require('webpack-dev-middleware')
+
+const webpack = require('webpack')
```
Can we please get rid of the excessive newlines? Nothing against lots of line breaks, but here it's a little excessive. We can group local and external dependencies though

gribnoysup replied:
Sure, will fix this! I was certain `s-c` eslint config wasn't allowing me to leave out the newlines between imports before... 🤔

philpl commented on sandbox/index.js:
```diff
@@ -0,0 +1,63 @@
+
+app.use(devMiddleware)
+
+app.use(livereloadMiddleware(compiler, server))
```
It'd be great if we could set up hot reloading (in a future PR please)

This would allow us to test hot reloading weirdness, which is unfortunately quite relevant right now

philpl commented on sandbox/index.js:
```diff
@@ -0,0 +1,63 @@
+  )
+
+  // Then we evaluate this script to get prerendered content
+  const { html, css } = eval(serverScript) // eslint-disable-line no-eval
```
I have some setup code lying around in one of my repos (which works like next.js) that uses the built-in hot reloader, dev middleware, and compiles both the server and the client in tandem without evaling.

I can probably implement this in a future separate PR, because I'm not comfortable having this approach here permanently, as some people might copy it for their own setups

gribnoysup replied:
If you want, I'm happy to help with this in a future PR, if you have some examples to look at 😉

gribnoysup commented:
@philpl I fixed all the nits. 

I moved all sandbox related stuff to the new section, but I didn't actually add anything about testing because I'm not familiar with how tests are structured in this repo, sorry.

I also fixed few minor styling errors in the `src` that @mxstbr was mentioning earlier

gribnoysup commented:
Tests are exiting with 0 but then the test run is finished with 1 😨

gribnoysup commented:
I can't understand why `travis-ci` test is not passing. All tests seem to pass, and the test script is exiting with `0` 😿

gribnoysup commented:
Reverted styling fixes to `constructors/constructWithOptions.js`, `utils/create-broadcast.js`, maybe this will make the tests pass again...

gribnoysup commented:
@mxstbr @philpl is everything ok? anything else missing from the PR?

mxstbr commented:
Wait why did you delete this?? :cry: I wanted this!

mxstbr commented:
@gribnoysup

philpl commented:
Yea, sorry, for ripping the ball here. I think if you restore the branch we can just merge it and keep on iterating :)

gribnoysup commented:
@mxstbr Sorry! There was no activity in this PR for a long time, so I thought that it is not wanted anymore and was planning on making a better version (especially without this hacky way of implementing a server-side live reload with `eval`). I can reopen this if you still want to go with this approach and then iterate on this somehow

gribnoysup commented:
@philpl @mxstbr sorry, didn't think that it is needed anymore, I'm reopening!

philpl commented:
@gribnoysup I've got a boilerplate with a webpack setup of mine that would be well suited here :wink: I can add it in a separate PR once we get this merged: https://github.com/philpl/micro-gql-react-ssr-boilerplate

Or you can pick it up? both works for me :smile:

Can you rebase the branch to resolve the yarn.lock conflicts please?

gribnoysup commented:
@philpl will be happy to help with that if you want!

resolved the `yarn.lock` conflict

philpl commented:
@mxstbr can you rereview and merge please? LGTM :)

gribnoysup commented:
Just want to note that there are still `eslint` spacing errors in the sandbox for 

https://github.com/styled-components/styled-components/blob/beea2b4ed450a9c06ce6b23ede0a6797c332fcef/src/constructors/constructWithOptions.js#L5-L7

and

https://github.com/styled-components/styled-components/blob/beea2b4ed450a9c06ce6b23ede0a6797c332fcef/src/utils/create-broadcast.js#L22

For some reason, they are not showing up when you just run `eslint` in the repo and last time I tried to fix them, TravisCI started failing for no reason.

I can fix them again, but I'm not sure if it will work this time. I still don't know why TravisCI was failing before...
